### PR TITLE
Add c10 cuda library.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#13900 Add c10 cuda library.**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13025313/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13910 Move AT_CUDA_CHECK to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13046201/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13912 Split out CUDAMultiStreamGuard from CUDAGuard&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13046500/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13920 Delete dependencies from CUDAStream; remove synchronize_with&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13047468/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14109 Move CUDAStreamInternals inside detail namespace.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13104694/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14110 Move stream functions from CUDAContext to CUDAStream&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13104693/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14191 Add hash functions for Stream, CUDAStream; fix Device hash function&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13119624/)

Add c10 cuda library.

Right now, this is not used by anything, and only tests if the CUDA
headers are available (and not, e.g., that linking works.)

Extra changes:
- cmake/public/cuda.cmake now is correctly include guarded, so you
  can include it multiple times without trouble.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D13025313](https://our.internmc.facebook.com/intern/diff/D13025313/)

### CODE REVIEW

File: c10/cuda/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Note: if you want to add ANY dependency to the c10 library, make sure you
+# check with the core PyTorch developers as the dependendency will be
+# transitively passed on to all libraries dependent on PyTorch.
+file(GLOB C10_CUDA_SRCS
+        *.cpp
+        *.cu
+        impl/*.cpp
+        impl/*.cu
+        )
+file(GLOB_RECURSE C10_CUDA_ALL_TEST_FILES test/*.cpp)
```
Comment (smessmer) [inline c10/cuda/CMakeLists.txt]: Let's not introduce a new glob, but immediately list the source files. We want to move other targets there as well.

File: c10/cuda/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# ---[ Test binaries.
+
+file(GLOB_RECURSE C10_CUDA_ALL_TEST_FILES *.cpp)
+if (BUILD_TEST)
+  foreach(test_src ${C10_CUDA_ALL_TEST_FILES})
+    get_filename_component(test_file_name ${test_src} NAME_WE)
+    set(test_name "c10_cuda_${test_file_name}")
+    add_executable(${test_name} "${test_src}")
+    target_link_libraries(${test_name} c10_cuda gtest_main)
+    add_test(NAME ${test_name} COMMAND $<TARGET_FILE:${test_name}>)
```
Comment (smessmer) [inline c10/cuda/test/CMakeLists.txt]: You're creating new executables here. Please make sure they're actually run in oss CI

**Review by smessmer — APPROVED** [review]
I don't see any changes in TARGETS files? Shouldn't you change them as well?

**Comment by ezyang** [comment]
TARGETS are in fbcode

File: c10/cuda/CMakeLists.txt (hunk 1 above)
Comment (ezyang) [reply c10/cuda/CMakeLists.txt]: Oh right. I had just copied this from c10/CMakeLists.txt and forgot that you're signed up to de-glob that cmake.

**Comment by ezyang** [comment]
Tests verified at: https://circleci.com/gh/pytorch/pytorch/262141?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Pull request closed without merging**